File: acmOtherMaterials/Ulm Local Contest/2005/solution/european.cpp
// Problem   European railroad tracks
// Algorithm backtracking
// Runtime   O(2^n * (n!)^2)
// Author    Adrian Kuegel
// Date      2005.05.27

#include <iostream>
#include <fstream>
#include <cassert>
#include <set>
using namespace std;

#define MAXN 8

int best;
set<int> solution;

void backtrack(int n, int tracks, set<int> &positions, int required[MAXN]) {
	if (tracks >= best)
		return;
	bool stop = true;
	// first check if one length is already covered by selected positions
	for (int i=1; i<n; i++) {
		if (!required[i])
			continue;
		stop = false;
		int old = required[i];
		for (set<int>::iterator it=positions.begin(); it!=positions.end(); it++) {
			if (positions.find(*it+old) != positions.end() || positions.find(*it-old) != positions.end()) {
				required[i] = 0;
				backtrack(n, tracks, positions, required);
				required[i] = old;
				return;
			}
		}
	}
	// if no required length is left, update best result and return
	if (stop) {
		best = tracks;
		solution = positions;
		return;
	}
	// if not, place another rail track
	// brute force over remaining lengths
	for (int i=1; i<n; i++) {
		if (!required[i])
			continue;
		int old = required[i];
		required[i] = 0;
		// place track in required distance of an old track
		for (set<int>::iterator it=positions.begin(); it!=positions.end(); it++) {
			positions.insert(*it+old);
			backtrack(n, tracks+1, positions, required);
			positions.erase(*it+old);
			positions.insert(*it-old);
			backtrack(n, tracks+1, positions, required);
			positions.erase(*it-old);
		}
		required[i] = old;
	}
}

int main() {
	int tc;
	ifstream in("european.in");
	in >> tc;
	for (int scen=1; scen<=tc; scen++) {
		cout << "Scenario #" << scen << endl;

		int n, required[MAXN];
		in >> n;
		assert(n >= 1 && n <= 8);
		solution.clear();
		solution.insert(0);
		for (int i=0; i<n; i++) {
			in >> required[i];
			assert(required[i]>=1000 && required[i]<=5000);
			solution.insert(required[i]);
		}
		set<int> positions;
		positions.insert(0);
		positions.insert(required[0]);
		required[0] = 0;
		best = n+1;
		backtrack(n, 2, positions, required);
		assert(best<=n+1 && best == solution.size());
		cout << best << ":";
		int offset = *(solution.begin());
		for (set<int>::iterator it = solution.begin(); it!=solution.end(); it++)
			cout << " " << *it-offset;
		cout << endl << endl;
	}
	return 0;
}

File: acmOtherMaterials/Ulm Local Contest/2005/solution/european.java
// Problem   European railroad tracks
// Algorithm backtracking
// Runtime   O(n^2 * n^(2*n))
// Author    Adrian Kuegel
// Date      2005.05.27

import java.util.*;
import java.io.*;

public class european {
	static StreamTokenizer in;
	static int nextInt() throws IOException {
		if (in.nextToken() != StreamTokenizer.TT_NUMBER)
			throw new RuntimeException("number expected");
		return (int) in.nval;
	}
	public static void main(String [] args) throws IOException {
		in = new StreamTokenizer(new BufferedReader(new FileReader("european.in")));
		int tc = nextInt();
		for(int scen = 1; scen <= tc; scen++) {
			int n = nextInt();
			int [] l = new int[n];
			for( int i = 0 ;i < n; i++ ) {
				l[i] = nextInt();
			}
			int [] left = new int[n];
			int [] right = new int[n];
			int bestcnt = n+1;
			try {
				for (bestcnt=1; bestcnt<=5; bestcnt++)
					backtrack(0,n,l,left,right,bestcnt);
			}catch(Exception e) {}
			System.out.println("Scenario #"+scen);
			System.out.print(bestcnt+":");
			for (int i=0; i<bestcnt; i++)
				for (int j=i+1; j<bestcnt; j++)
					if (solution[j] < solution[i]) {
						int t = solution[j];
						solution[j] = solution[i];
						solution[i] = t;
					}
			for (int i=0; i<bestcnt; i++)
				System.out.print(" "+(solution[i]-solution[0]));
			System.out.println("\n");
		}
	}
	static int[] solution;
	// this function tries to find a solution using "rails" rails
	// if it finds such a solution, it throws an exception
	static void backtrack( int act , int n, int [] l, int [] left, int [] right, int rails) throws Exception {
		if( act == n ) {
			// position of the rails
			int [] tmp = new int[rails];
			int oo = 1000000;
			// initialize to undefined
			for( int i = 0; i < rails; i++ )
				tmp[i] = oo;
			// one rail gets position 0
			tmp[0] = 0;
			for( int i = 0; i < n; i++ ) {
				int cnt = 0;
				for( int j = 0; j < n; j++ ) {
					// if both rails of a gauge have a position, check that the distance is correct
					if( tmp[left[j]] != oo && tmp[right[j]] != oo ) {
						if( tmp[right[j]] != tmp[left[j]] +l[j] ) return;
						cnt++;
					}
					// if the left position is known, assign the right position
					else if( tmp[left[j]] != oo ) {
						tmp[right[j]] = tmp[left[j]] + l[j];
						cnt++;
					}
					// if the right position is known, assign the left position
					else if( tmp[right[j]] != oo ) {
						tmp[left[j]] = tmp[right[j]] - l[j];
						cnt++;
					}
				}
				// all gauges have been found
				if( cnt == n ) {
					solution = tmp;
					throw new Exception("done");
				}
			}
			return;
		}
		// first, assign gauges to rails without assigning any position to the rails
		for( int j = 0; j < rails; j++ )
			for( int k = j+1; k < rails ;k++ ) {
				left[act] = j;
				right[act] = k;
				backtrack( act+1, n, l, left, right, rails);
			}
	}
}